File: usage-daemon/src/providers/claude.js
// Claude Code usage provider plugin — descriptor interface (HANDOFF-14).
//
// Reads the OAuth usage endpoint with the same accessToken the `claude` CLI
// itself uses (~/.claude/.credentials.json). READ-ONLY: this daemon never
// writes that file and never refreshes the token — expired/missing token
// just reports auth_expired (see AuthExpiredError below); the CLI or `claude
// login` remain the only things that mutate it.
//
// parse() is a PURE function of the raw JSON text so it unit-tests against
// the vendored fixture (test/fixtures/claude-usage.json) with no network/fs.
// fetch() adds the credentials read + HTTP fetch + auth-expiry detection
// around it. Field mapping ported verbatim from
// claude-usage-extension/extension.js (the live extractor is the spec).

import { promises as fs } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';

const execFileP = promisify(execFile);

export const SESSION_COLOR = '#E69F00'; // Okabe-Ito orange (suite-wide)
export const WEEKLY_COLOR = '#56B4E9';  // Okabe-Ito blue
export const EXTRA_COLOR = '#009E73';   // Okabe-Ito green — extra/overage credits

export const ID = 'claude';
export const LABEL = 'Claude Code';

const API_URL = 'https://api.anthropic.com/api/oauth/usage';
// Anthropic buckets requests whose claude-code/<version> UA is too far behind
// the current CLI into aggressive 429s — see anthropics/claude-code#31021,
// #30930, #31637. Confirmed live 2026-07-31: UA claude-code/2.1.0 -> 429,
// claude-code/2.1.220 -> 200 against the same token. So the version MUST track
// the real CLI, not a frozen literal.
//
// Resolution order (see resolveUserAgentVersion): explicit config
// (`[providers.claude] user_agent_version`) > auto-detected `claude --version`
// (cached once at first poll) > this hardcoded fallback. Bump the fallback when
// you notice 429s creep back and can't rely on a local CLI to auto-detect.
const FALLBACK_CLAUDE_VERSION = '2.1.220';
const CLAUDE_UA_PREFIX = 'claude-code/';

// Extract a semver-ish "1.2.3" from `claude --version` output
// ("2.1.220 (Claude Code)") or a user-supplied version string.
function extractVersion(s) {
  const m = String(s ?? '').match(/(\d+\.\d+\.\d+)/);
  return m ? m[1] : null;
}

// Detect the installed CLI version once; cache the promise so concurrent polls
// share a single `claude --version` spawn. Never throws — resolves null on any
// failure (not installed, PATH miss, timeout) so the caller falls back.
let detectPromise = null;
function detectCliVersion() {
  if (!detectPromise) {
    detectPromise = execFileP('claude', ['--version'], { timeout: 5000 })
      .then(({ stdout }) => extractVersion(stdout))
      .catch(() => null);
  }
  return detectPromise;
}

export class AuthExpiredError extends Error {
  constructor(msg = 'Claude Code token missing or expired') {
    super(msg);
    this.code = 'auth_expired';
  }
}

export class RateLimitedError extends Error {
  constructor(retryAfter = null) {
    super('rate_limited');
    this.code = 'rate_limited';
    this.retryAfter = retryAfter;
  }
}

function defaultCredentialsPath() {
  const configDir = process.env.CLAUDE_CONFIG_DIR || path.join(os.homedir(), '.claude');
  return path.join(configDir, '.credentials.json');
}

// Pure function of the raw API JSON text — no fs/network. Throws
// AuthExpiredError if the payload doesn't look like a usage response
// (mirrors ollama's parse throwing on a logged-out page).
export function parse(raw) {
  let data;
  try {
    data = JSON.parse(raw);
  } catch {
    throw new AuthExpiredError('unparseable usage response');
  }
  if (!data?.five_hour || !data?.seven_day) throw new AuthExpiredError();

  const windows = [
    {
      id: 'session',
      label: '5h',
      letter: '5h',
      pct: data.five_hour.utilization ?? null,
      resets_at: data.five_hour.resets_at ?? null,
      color: SESSION_COLOR,
      will_deplete: false,
    },
    {
      id: 'weekly',
      label: '7d',
      letter: 'Wk',
      pct: data.seven_day.utilization ?? null,
      resets_at: data.seven_day.resets_at ?? null,
      color: WEEKLY_COLOR,
      will_deplete: false,
    },
  ];

  // Extra usage ("usage credits") — pay-as-you-go spend that kicks in once a
  // plan limit is hit. Only surfaced when the feature is enabled on the
  // account (otherwise the block is present-but-off or null). Credits are
  // integer minor units scaled by decimal_places (e.g. 1361 @ dp=2 -> $13.61)
  // against monthly_limit (2000 @ dp=2 -> $20.00). Resets on the 1st of the
  // month (the API gives no reset date, so we derive next-month-start).
  const eu = data.extra_usage;
  if (eu && eu.is_enabled) {
    const dp = typeof eu.decimal_places === 'number' ? eu.decimal_places : 2;
    const div = 10 ** dp;
    const used = typeof eu.used_credits === 'number' ? eu.used_credits / div : null;
    const cap = typeof eu.monthly_limit === 'number' ? eu.monthly_limit / div : null;
    // Prefer the API's own utilization; fall back to used/cap when it's absent
    // or null (the endpoint stopped sending `utilization`, which left the bar
    // blank even though used+cap were present).
    const pct =
      typeof eu.utilization === 'number'
        ? eu.utilization
        : cap && cap > 0 && used != null
          ? (100 * used) / cap
          : null;
    windows.push({
      id: 'extra_usage',
      label: 'Usage Credits',
      letter: 'Cr',
      pct,
      used,
      cap,
      unit: eu.currency || 'USD',
      resets_at: nextMonthStart(),
      color: EXTRA_COLOR,
      will_deplete: false,
    });
  }

  return { tier: null, windows, segments: [] };
}

// First of next month, 00:00 UTC — the extra-usage monthly cap's reset
// boundary (the usage API omits a reset date; the account dashboard shows
// "Resets <1st of next month>").
export function nextMonthStart(from = new Date()) {
  const y = from.getUTCFullYear();
  const m = from.getUTCMonth();
  return new Date(Date.UTC(y, m + 1, 1, 0, 0, 0)).toISOString().replace('.000', '');
}

function createProvider() {
  let credentialsPath = defaultCredentialsPath();
  let lastTokenExpiresAt = null; // surfaced via meta() — set on each successful credentials read
  let configuredVersion = null;  // explicit override from config, if any
  let resolvedUserAgent = null;  // cached final UA string once resolved

  // Resolve the claude-code/<version> UA once: config override > detected CLI
  // version > hardcoded fallback. Cached after the first successful resolve.
  async function resolveUserAgent() {
    if (resolvedUserAgent) return resolvedUserAgent;
    const version =
      configuredVersion || (await detectCliVersion()) || FALLBACK_CLAUDE_VERSION;
    resolvedUserAgent = CLAUDE_UA_PREFIX + version;
    return resolvedUserAgent;
  }

  return {
    id: ID,
    label: LABEL,
    auth: { kind: 'oauth-file' },

    config() {
      return {
        id: ID,
        label: LABEL,
        usageUrl: 'https://claude.ai/settings/usage',
        // path/relogin let a client explain an auth_expired instead of just
        // showing it: this token is read straight from the Claude CLI's own
        // credentials file and never refreshed here, so re-login IS the fix.
        auth: { kind: 'oauth-file', path: credentialsPath, relogin: 'claude  (then /login)' },
        windows: [
          { id: 'session', label: '5h', color: SESSION_COLOR },
          { id: 'weekly', label: '7d', color: WEEKLY_COLOR },
          { id: 'extra_usage', label: 'Usage Credits', color: EXTRA_COLOR },
        ],
        tiers: [],
      };
    },

    configure(cfg = {}) {
      const p = cfg.credentialsPath || cfg.credentials_path;
      if (p) credentialsPath = p;
      const v = extractVersion(cfg.user_agent_version || cfg.userAgentVersion);
      if (v) {
        configuredVersion = v;
        resolvedUserAgent = null; // force re-resolve with the new override
      }
    },

    async setAuth(payload) {
      await fs.mkdir(path.dirname(credentialsPath), { recursive: true });
      await fs.writeFile(credentialsPath, payload + '\n', { mode: 0o600 });
    },

    async fetch() {
      let accessToken;
      try {
        const raw = await fs.readFile(credentialsPath, 'utf8');
        const json = JSON.parse(raw);
        accessToken = json?.claudeAiOauth?.accessToken;
        lastTokenExpiresAt = json?.claudeAiOauth?.expiresAt ?? null;
      } catch {
        throw new AuthExpiredError('no Claude Code credentials file found');
      }
      if (!accessToken) throw new AuthExpiredError('no accessToken in credentials file');

      const res = await fetch(API_URL, {
        headers: {
          Authorization: `Bearer ${accessToken}`,
          'anthropic-beta': 'oauth-2025-04-20',
          'User-Agent': await resolveUserAgent(),
        },
      });
      if (res.status === 401) throw new AuthExpiredError();
      if (res.status === 429) {
        const retryAfter = Number(res.headers.get('retry-after')) || null;
        throw new RateLimitedError(retryAfter);
      }
      if (!res.ok) throw new Error(`api.anthropic.com HTTP ${res.status}`);
      return res.text();
    },

    intervalSeconds() {
      return 300;
    },

    // Optional generic hook (runner merges this into the snapshot if
    // present) — surfaces facts that live outside the parsed payload, here
    // the credentials file's own expiry, not part of the usage API response.
    meta() {
      return { token_expires_at: lastTokenExpiresAt };
    },

    parse,
  };
}

export { createProvider };

File: usage-daemon-v3/usage_daemon/providers/claude.py
"""Claude Code usage provider plugin (port of src/providers/claude.js).

Reads the OAuth usage endpoint with the same accessToken the `claude` CLI uses
(`~/.claude/.credentials.json`). In practice this host may also carry a raw
bearer token in that path; accept either shape so a fresh v3 process can use
what the live environment already has on disk. READ-ONLY: never mutates the
token — an expired/missing one just reports auth_expired; `claude login` is the
mutator.

The claude-code/<version> User-Agent gates real 429s: Anthropic buckets UAs too
far behind the current CLI. Resolution: config override > detected
`claude --version` > hardcoded fallback (FALLBACK_CLAUDE_VERSION).
"""

from __future__ import annotations

import asyncio
import datetime as _dt
import json
import os
import re
import shutil
from pathlib import Path

from ..errors import AuthExpiredError, RateLimitedError
from ..httputil import create_client

SESSION_COLOR = "#E69F00"
WEEKLY_COLOR = "#56B4E9"
EXTRA_COLOR = "#009E73"

ID = "claude"
LABEL = "Claude Code"

API_URL = "https://api.anthropic.com/api/oauth/usage"
FALLBACK_CLAUDE_VERSION = "2.1.220"
CLAUDE_UA_PREFIX = "claude-code/"


def extract_version(s) -> str | None:
    m = re.search(r"(\d+\.\d+\.\d+)", s or "")
    return m.group(1) if m else None


def _default_credentials_path() -> str:
    cfg = os.environ.get("CLAUDE_CONFIG_DIR") or os.path.join(
        os.path.expanduser("~"), ".claude"
    )
    return os.path.join(cfg, ".credentials.json")


def next_month_start(from_dt=None) -> str:
    """First of next month, 00:00 UTC — extra-usage monthly cap reset boundary."""
    d = from_dt or _dt.datetime.now(_dt.timezone.utc)
    if d.tzinfo is None:
        d = d.replace(tzinfo=_dt.timezone.utc)
    y, m = d.year, d.month
    nm = _dt.datetime(
        y + (1 if m == 12 else 0),
        1 if m == 12 else m + 1,
        1,
        tzinfo=_dt.timezone.utc,
    )
    return nm.isoformat()


def parse(raw) -> dict:
    """Pure function of the raw API JSON text -> {tier, windows, segments}."""
    try:
        data = json.loads(raw)
    except Exception:
        raise AuthExpiredError("unparseable usage response")
    if not data.get("five_hour") or not data.get("seven_day"):
        raise AuthExpiredError()

    five = data["five_hour"]
    seven = data["seven_day"]
    windows = [
        {
            "id": "session",
            "label": "5h",
            "letter": "5h",
            "pct": five.get("utilization"),
            "resets_at": five.get("resets_at"),
            "color": SESSION_COLOR,
            "will_deplete": False,
        },
        {
            "id": "weekly",
            "label": "7d",
            "letter": "Wk",
            "pct": seven.get("utilization"),
            "resets_at": seven.get("resets_at"),
            "color": WEEKLY_COLOR,
            "will_deplete": False,
        },
    ]

    eu = data.get("extra_usage")
    if eu and eu.get("is_enabled"):
        dp = eu["decimal_places"] if isinstance(eu.get("decimal_places"), (int, float)) else 2
        div = 10 ** int(dp)
        used = (
            eu["used_credits"] / div
            if isinstance(eu.get("used_credits"), (int, float))
            else None
        )
        cap = (
            eu["monthly_limit"] / div
            if isinstance(eu.get("monthly_limit"), (int, float))
            else None
        )
        if isinstance(eu.get("utilization"), (int, float)):
            pct = eu["utilization"]
        elif cap and cap > 0 and used is not None:
            pct = (100 * used) / cap
        else:
            pct = None
        windows.append({
            "id": "extra_usage",
            "label": "Usage Credits",
            "letter": "Cr",
            "pct": pct,
            "used": used,
            "cap": cap,
            "unit": eu.get("currency") or "USD",
            "resets_at": next_month_start(),
            "color": EXTRA_COLOR,
            "will_deplete": False,
        })

    return {"tier": None, "windows": windows, "segments": []}


async def _detect_cli_version_once(cache: dict) -> str | None:
    """Detect the installed CLI version once; cache the result for concurrent polls."""
    if "done" in cache:
        return cache.get("value")
    cache["done"] = True
    cli = shutil.which("claude")
    if cli is None:
        cache["value"] = None
        return None
    proc = await asyncio.create_subprocess_exec(
        cli,
        "--version",
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.DEVNULL,
    )
    try:
        out, _ = await asyncio.wait_for(proc.communicate(), timeout=5)
    except Exception:
        cache["value"] = None
        return None
    cache["value"] = extract_version(out.decode("utf-8", "replace"))
    return cache["value"]


def create_provider(credentials_path: str | None = None, client=None):
    st = {
        "credentialsPath": credentials_path or _default_credentials_path(),
        "lastTokenExpiresAt": None,
        "configuredVersion": None,
        "resolvedUserAgent": None,
        "client": client,
        "detectCache": {},
    }

    def config():
        return {
            "id": ID,
            "label": LABEL,
            "usageUrl": "https://claude.ai/settings/usage",
            "auth": {
                "kind": "oauth-file",
                "path": st["credentialsPath"],
                "relogin": "claude  (then /login)",
            },
            "windows": [
                {"id": "session", "label": "5h", "color": SESSION_COLOR},
                {"id": "weekly", "label": "7d", "color": WEEKLY_COLOR},
                {"id": "extra_usage", "label": "Usage Credits", "color": EXTRA_COLOR},
            ],
            "tiers": [],
        }

    def configure(cfg: dict | None = None):
        cfg = cfg or {}
        p = cfg.get("credentialsPath") or cfg.get("credentials_path")
        if p:
            st["credentialsPath"] = p
        v = extract_version(cfg.get("user_agent_version") or cfg.get("userAgentVersion"))
        if v:
            st["configuredVersion"] = v
            st["resolvedUserAgent"] = None

    async def set_auth(payload: str) -> None:
        path = Path(st["credentialsPath"])
        await asyncio.to_thread(path.parent.mkdir, exist_ok=True, parents=True)

        def _write():
            path.write_text(payload + "\n", encoding="utf-8")
            try:
                os.chmod(path, 0o600)
            except OSError:
                pass

        await asyncio.to_thread(_write)

    async def _resolve_user_agent() -> str:
        if st["resolvedUserAgent"]:
            return st["resolvedUserAgent"]
        v = (
            st["configuredVersion"]
            or await _detect_cli_version_once(st["detectCache"])
            or FALLBACK_CLAUDE_VERSION
        )
        st["resolvedUserAgent"] = CLAUDE_UA_PREFIX + v
        return st["resolvedUserAgent"]

    async def fetch() -> str:
        def _read():
            try:
                return Path(st["credentialsPath"]).read_text(encoding="utf-8")
            except OSError:
                return None

        raw = await asyncio.to_thread(_read)
        if raw is None:
            raise AuthExpiredError("no Claude Code credentials file found")
        text = raw.strip()
        access_token = None
        st["lastTokenExpiresAt"] = None
        try:
            j = json.loads(text)
        except Exception:
            # Practical compatibility: some local setups store the bearer token
            # directly in this path instead of the CLI's JSON envelope.
            access_token = text or None
        else:
            access_token = (j or {}).get("claudeAiOauth", {}).get("accessToken")
            st["lastTokenExpiresAt"] = (j or {}).get("claudeAiOauth", {}).get("expiresAt")
        if not access_token:
            raise AuthExpiredError("no accessToken in credentials file")

        c = st["client"] or create_client()
        res = await c.get(
            API_URL,
            headers={
                "Authorization": f"Bearer {access_token}",
                "anthropic-beta": "oauth-2025-04-20",
                "User-Agent": await _resolve_user_agent(),
            },
        )
        if res.status_code in (401, 403):
            detail = None
            try:
                err = res.json().get("error") or {}
                detail = err.get("message")
                scopes = ((err.get("details") or {}).get("required_scopes"))
                if isinstance(scopes, list) and scopes:
                    detail = f"{detail or 'OAuth token rejected'} (required scopes: {', '.join(map(str, scopes))})"
            except Exception:
                detail = None
            raise AuthExpiredError(detail or "Claude Code token missing, expired, or insufficiently scoped")
        if res.status_code == 429:
            ra = res.headers.get("retry-after")
            raise RateLimitedError(int(ra) if ra and ra.isdigit() else None)
        if res.status_code >= 400:
            raise RuntimeError(f"api.anthropic.com HTTP {res.status_code}")
        return res.text

    def interval_seconds():
        return 300

    def meta():
        return {"token_expires_at": st["lastTokenExpiresAt"]}

    return {
        "id": ID,
        "label": LABEL,
        "auth": {"kind": "oauth-file"},
        "config": config,
        "configure": configure,
        "set_auth": set_auth,
        "fetch": fetch,
        "interval_seconds": interval_seconds,
        "meta": meta,
        "parse": parse,
    }